### DISCUSSION TITLE
Setting variable location over time +  saying gdb to ignore calls

### DISCUSSION THREAD

**Discussion by Cr0a3** [gimli-rs/gimli#748] [Q&A] [answered]
Hi,
I recently started learning gimli.
I made a simple programm: https://gist.github.com/Cr0a3/33eb2ad87ba7f807f5473e2851264f7d which is a combinantion of the `simple_write` example and tried out things from me.

**First question:**
I set my variable to a location here: https://gist.github.com/Cr0a3/33eb2ad87ba7f807f5473e2851264f7d#file-main-rs-L151-L154
and override it here: https://gist.github.com/Cr0a3/33eb2ad87ba7f807f5473e2851264f7d#file-main-rs-L162-L165
How can i include that the variable location changes on a specific adress (like my line programms)?

I am searching something like this:
```
dwarf.unit.get_mut(x).set(
    DW_AT_location,
    AttributeValue::Exprloc(rax),
    /*where*/set2,
);
```

**Second question:**
How can i include in the debugging information generated by gimli that gdb should ignore calls and just step over them?

Bye

**Comment by philipc** [comment, marked as answer]
Use [AttributeValue::LocationListRef](https://docs.rs/gimli/latest/gimli/write/enum.AttributeValue.html#variant.LocationListRef) to refer to a [LocationList](https://docs.rs/gimli/latest/gimli/write/struct.LocationList.html).

The debugging information can't control whether gdb ignores calls. Instead, that is a result of the commands the user inputs to gdb.

**Reply by Cr0a3** [reply]
Thank you,
I now added following:
```
    let mut rdi = Expression::new();
    rdi.op_reg(gimli::Register(5));
    let mut rax = Expression::new();
    rax.op_reg(gimli::Register(0));

    let mut locs = vec![];
    locs.push(
        Location::StartEnd { 
            begin: Address::Symbol { symbol: 0, addend: set1 as i64 }, 
            end: Address::Symbol { symbol: 0, addend: set2 as i64 }, 
            data: rdi
        }
    );

    locs.push(
        Location::StartEnd { 
            begin: Address::Symbol { symbol: 0, addend: set2 as i64 }, 
            end: Address::Symbol { symbol: 0, addend: main_size as i64 }, 
            data: rax
        }
    );

    let locs_id = dwarf.unit.locations.add(LocationList(locs));

    dwarf.unit.get_mut(x).set(
        DW_AT_location,
        AttributeValue::LocationListRef(locs_id),
    ); 
```
The dwarfdump shows it correctly:
```
0x0000000c: DW_TAG_compile_unit
              DW_AT_producer    ("test")
              DW_AT_language    (DW_LANG_C)
              DW_AT_name        ("hello.c")
              DW_AT_comp_dir    ("./tmp")
              DW_AT_low_pc      (0x0000000000000000)
              DW_AT_ranges      (0x0000000c
                 [0x0000000000000000, 0x0000000000000025))
              DW_AT_stmt_list   (0x00000000)

0x00000031:   DW_TAG_variable
                DW_AT_name      ("x")
                DW_AT_location  (0x0000000c:
                   [0x0000000000000010, 0x0000000000000017): DW_OP_reg5 RDI
                   [0x0000000000000017, 0x0000000000000025): DW_OP_reg0 RAX)

0x00000038:   DW_TAG_subprogram
                DW_AT_external  (0x01)
                DW_AT_name      ("main")
                DW_AT_decl_file ("./tmp/hello.c")
                DW_AT_decl_line (3)
                DW_AT_decl_column       (0)
                DW_AT_low_pc    (0x0000000000000000)
                DW_AT_high_pc   (0x0000000000000025)

0x0000004b:   NULL
```
I veryfied the locations with `llvm-objdump` and they were correct (for the assign of the variable).
But if i go into gdb and then debug the pogram to test if the moved variable location worked:
```
(gdb) p x
$3 = 13
(gdb)  info register rax                                                                                                                                    
rax            0xd                 13
(gdb) s
(gdb) p x
$4 = 2
...
```
Which means gdb thinks that the variable is always in rax. 
Bye

**Reply by bjorn3** [reply]
I suspect you meant to put `DW_TAG_variable` inside `DW_TAG_subprogram`. As is you have defined a global variable. Gdb may assume that these have a fixed location.
